File: Utils/Trees/Tree.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

    def __str__(self) -> str:
        return f'{self.val}'

class Tree:
    def __init__(self):
        self.root = None

    # Return the in-order traversal of the tree as a string
    def __str__(self):
        values = []
        self._level_order(self.root, values)
        return " -> ".join(map(str, values))

    # In-order traversale for print statement
    def _level_order(self, node, values):
        if not node:
            return
        
        queue = [node]

        while queue:
            curr = queue.pop(0)
            values.append(curr.val)

            if curr.left:
                queue.append(curr.left)

            if curr.right:
                queue.append(curr.right)


    def add(self, value):
        if not self.root:
            self.root = TreeNode(value)
            return

        queue = [self.root]

        while queue:
            current = queue.pop(0)

            if not current.left:
                current.left = TreeNode(value)
                return
            
            if not current.right:
                current.right = TreeNode(value)
                return
            
            queue.append(current.left)
            queue.append(current.right)
        

File: Utils/Trees/Tree.ts
export class TreeNode {
    val: number;
    left: TreeNode | null;
    right: TreeNode | null;

    constructor(val: number) {
        this.val = val;
        this.left = null;
        this.right = null;
    }

    toString(): string {
        return `${this.val}`;
    }
}

export class Tree {
    root: TreeNode | null;
    constructor() {
        this.root = null;
    }

    // Return the in-order traversal of the tree as a string
    toString() {
        let values: number[] = [];
        this._level_order(this.root, values);
        return values.join(" -> ");
    }
    // In-order traversale for print statement
    _level_order(node: TreeNode | null, values: number[]) {
        if (!node) return;

        let queue = [node];

        while (queue.length > 0) {
            let curr = queue.shift()!;
            values.push(curr.val);

            if (curr.left) {
                queue.push(curr.left);
            }

            if (curr.right) {
                queue.push(curr.right);
            }
        }
    }

    add(value: number): void {
        if (!this.root) {
            this.root = new TreeNode(value);
            return;
        }

        let queue = [this.root];

        while (queue.length > 0) {
            let current = queue.shift()!;

            if (!current.left) {
                current.left = new TreeNode(value);
                return;
            }

            if (!current.right) {
                current.right = new TreeNode(value);
                return;
            }

            queue.push(current.left);
            queue.push(current.right);
        }
    }
}
